File: src/cli/plan.test.ts
import { describe, expect, test } from "bun:test";

import { MAINSTREAM_TLDS, createQueryPlan } from "@/cli/plan";

describe("query plan", () => {
  test("keeps explicit domain exact", () => {
    const plan = createQueryPlan(["murk.ink"]);

    expect(plan.groups[0]?.suggested).toBeFalse();
    expect(plan.groups[0]?.domains).toEqual(["murk.ink"]);
    expect(plan.lookupDomains).toEqual(["murk.ink"]);
  });

  test("expands bare domain to mainstream tld list", () => {
    const plan = createQueryPlan(["murk"]);
    const expected = MAINSTREAM_TLDS.map((tld) => `murk.${tld}`);

    expect(plan.groups[0]?.suggested).toBeTrue();
    expect(plan.groups[0]?.domains).toEqual(expected);
    expect(plan.lookupDomains).toEqual(expected);
  });
});

File: internal/cli/plan_test.go
package cli

import (
	"testing"
)

func TestExplicitDomainExact(t *testing.T) {
	groups, lookupDomains := createQueryPlan([]string{"murk.ink"})

	if len(groups) != 1 {
		t.Fatalf("expected 1 group, got %d", len(groups))
	}
	if groups[0].Suggested {
		t.Error("expected suggested to be false")
	}
	if len(groups[0].Domains) != 1 || groups[0].Domains[0] != "murk.ink" {
		t.Errorf("unexpected domains: %v", groups[0].Domains)
	}
	if len(lookupDomains) != 1 || lookupDomains[0] != "murk.ink" {
		t.Errorf("unexpected lookup domains: %v", lookupDomains)
	}
}

func TestBareLabelExpansion(t *testing.T) {
	groups, lookupDomains := createQueryPlan([]string{"murk"})

	if len(groups) != 1 {
		t.Fatalf("expected 1 group, got %d", len(groups))
	}
	if !groups[0].Suggested {
		t.Error("expected suggested to be true")
	}

	expected := make([]string, len(mainstreamTLDs))
	for i, tld := range mainstreamTLDs {
		expected[i] = "murk." + tld
	}

	if len(groups[0].Domains) != len(expected) {
		t.Fatalf("expected %d domains, got %d", len(expected), len(groups[0].Domains))
	}
	for i, d := range groups[0].Domains {
		if d != expected[i] {
			t.Errorf("domain[%d] = %q, want %q", i, d, expected[i])
		}
	}
	if len(lookupDomains) != len(expected) {
		t.Errorf("expected %d lookup domains, got %d", len(expected), len(lookupDomains))
	}
}
